### DISCUSSION TITLE
mandatory fields are not being validated

### DISCUSSION THREAD

**Discussion by roylisto** [cue-lang/cue#1949] [Q&A]
Hi, I use unify() before validate()

```
package main

import (
	"fmt"

	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/cue/errors"
	yamlCue "cuelang.org/go/encoding/yaml"
)

var yamlData = []byte(`
field1: test
field2: test
`)

func main() {
	ctx := cuecontext.New()

	v := ctx.CompileString(`
        field1: "hello"
        field2: "hello"
        `)

	yf, _ := yamlCue.Extract("test.yaml", yamlData)
	yv := ctx.BuildFile(yf)

	v = v.Unify(yv)
	err := v.Validate()
	if err != nil {
		fmt.Println(errors.Details(err, nil))
	}
}
```

by using unify() and validate() if I omit field1 or field2, it will not throw any error, I am expecting those to be error
how can I fix this? 

update:

here is my code for trial
```
package main

import (
	"fmt"

	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/cue/errors"
	yamlCue "cuelang.org/go/encoding/yaml"
)

var yamlData = []byte(`
field3: test
`)

func main() {
	ctx := cuecontext.New()

	v := ctx.CompileString(`
        field1: "hello"
        field2: "hello"
        `)

	yf, err := yamlCue.Extract("test.yaml", yamlData)
	fmt.Println(err)
	yv := ctx.BuildFile(yf)

	v = v.Unify(yv)
	err = v.Validate()
	if err != nil {
		fmt.Println(errors.Details(err, nil))
	}
}
```
I am expecting an error saying the field1 and field2 is required
thank you

**Comment by verdverm** [comment]
You may need to check `v.Err() != nil` after `v.Unify()`

**Reply by roylisto** [reply]
error is `nil`

**Comment by mpvl** [comment]
This works for me, after tweaking the YAML a bit. You don't need to check the error after `Unify`, as `Validate` will get that too.

However, I did need to adjust the YAML a bit. Keep in mind that, unlike CUE, YAML is extremely sensitive to whitespace. So be off by one tab, and the parsing fails. If it fails, then building the resulting "null" value has the effect of not unifying anything and everything next will succeed.

So, be sure to catch possible errors in the YAML code by checking the error returned by `yaml.Extract`.

**Reply by roylisto** [reply]
I can't see any error thrown from Exract() function, here is my code
```
package main

import (
	"fmt"

	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/cue/errors"
	yamlCue "cuelang.org/go/encoding/yaml"
)

var yamlData = []byte(`
field3: test
`)

func main() {
	ctx := cuecontext.New()

	v := ctx.CompileString(`
        field1: "hello"
        field2: "hello"
        `)

	yf, err := yamlCue.Extract("test.yaml", yamlData)
	fmt.Println(err)
	yv := ctx.BuildFile(yf)

	v = v.Unify(yv)
	err = v.Validate()
	if err != nil {
		fmt.Println(errors.Details(err, nil))
	}
}
```

**Reply by roylisto** [reply]
I am expecting an error saying the field1 and field2 is required

**Reply by mpvl** [reply]
Ah, I copied the first example and my editor indented it.

Right now there is no good way to require a "concrete" field. This is one of several thing that is related to the required fields proposal that was just published: https://github.com/cue-lang/cue/discussions/1951.

**Reply by roylisto** [reply]
oh okay, so no way to check the required field for now right?

**Reply by roylisto** [reply]
or can I know what field is missing from the yaml file? @mpvl

**Reply by myitcv** [reply]
If you want an error that highlights `field1` and `field2` are missing from the YAML, then you cannot do that today. That is covered by the implementation that will follow #1951.

**Reply by roylisto** [reply]
so now all the fields are optional then?

**Reply by myitcv** [reply]
No, regular fields today (combined with the requiring of things being concrete) makes things mandatory. e.g:

For example, given the schema `#Person` and value `p`:

```
#Person: {
    name: string
}

p: #Person
```

This will error when exported to concrete configuration (e.g. JSON) because `name` has not been specified.

In effect this requires the user (the author of `p`) to define `name`:

```
p: #Person & {
     name: "pete"
}
```

What is _not_ possible is to require the user to specify a field which already has a concrete value:

```
#Person: {
    type!: "human"
    name!: string
}
```

**Reply by roylisto** [reply]
Oh okay, so it's not possible right now to make `name` to be mandatory by using a predefined value like this?
```
#Person: {
    name:  'hello'
}

p: #Person
```

**Reply by roylisto** [reply]
I found a hacky way by using regex like this, for now, it can detect if the fields are defined or not in the YAML file
```
field1: =~"^hello$"
field2: =~"^world$"
```

**Reply by myitcv** [reply]
> Oh okay, so it's not possible right now to make `name` to be mandatory by using a predefined value like this?
> 
> ```
> #Person: {
>     name:  'hello'
> }
> 
> p: #Person
> ```

To be precise, what's not possible today is to require that the field `name` be defined by `p`. To make this more explicit, imagine if `p` were instead some input JSON you wanted to validate. Considering the way CUE behaves today, the JSON input `{}` satisfies `#Person`:

```
exec cue export -d '#Person' input.json schema.cue
cmp stdout stdout.golden

-- input.json --
{}
-- schema.cue --
#Person: {
	name: "hello"
}
-- stdout.golden --
{
    "name": "hello"
}
```

With required fields we could instead require that `name` be defined, in which case the behaviour would be something like this:

```
! exec cue export -d '#Person' input.json schema.cue
stderr 'field name is required'

-- input.json --
{}
-- schema.cue --
#Person: {
	name!: "hello"
}
```

Notice: the expectation here is that the `cue export` command will _fail_, and that the error message will include something like `field name is required`. To fix this we correct the input:

```
exec cue export -d '#Person' input.json schema.cue
cmp stdout stdout.golden

-- input.json --
{
    "name": "hello"
}
-- schema.cue --
#Person: {
	name!: "hello"
}
-- stdout.golden --
{
    "name": "hello"
}
```

**Reply by roylisto** [reply]
thanks, @myitcv for the explanation! Interesting feature!
